sklearn/cluster/_hdbscan/hdbscan.py

```diff
@@ -0,0 +1,718 @@
+                self._single_linkage_tree_, internal_to_raw, outliers
+            )
+            new_labels = np.full(self._raw_data.shape[0], -1)
+            new_labels[finite_index] = self.labels_
+            self.labels_ = new_labels
+
+            new_probabilities = np.zeros(self._raw_data.shape[0])
+            new_probabilities[finite_index] = self.probabilities_
+            self.probabilities_ = new_probabilities
+
+        return self
+
+    def fit_predict(self, X, y=None):
+        """Perform clustering on X and return cluster labels.
+
+        Parameters
+        ----------
+        X : array or sparse (CSR) matrix of shape (n_samples, n_features), or \
+                array of shape (n_samples, n_samples)
+            A feature array, or array of distances between samples if
+            `metric='precomputed'`.
+
+        y : Ignored
+            Ignored.
+
+        Returns
+        -------
+        y : ndarray, shape (n_samples, )
+            Cluster labels.
+        """
+        self.fit(X)
+        return self.labels_
+
+    def weighted_cluster_center(self, cluster_id, mode="centroid"):
+        """
+        Provide an approximate representative point for a given cluster.
+
+        Parameters
+        ----------
+        cluster_id : int
+            The id of the cluster to compute a centroid for.
+
+        mode : str, default="centroid"
+            The mode to use when providing the cluster center. The options are:
+            - "centroid" which calculates the center by taking the weighted
+              average of their positions. Note that the algorithm assumes a
+              euclidean metric and does not guarantee that the output will be
+              an observed data point.
+            - "medoid" which calculates the center by taking the point in the
+              fitted data which minimizes the distance to all other points in
+              the cluster. This is slower than "centroid" since it requires
+              computing additional pairwise distances between points of the
+              same cluster but guarantees the output is an observed data point.
+
+        Returns
+        -------
+        centroid : array of shape (n_features,)
+            A representative centroid for cluster `cluster_id`.
+        """
+        if not hasattr(self, "labels_"):
+            raise AttributeError("Model has not been fit to data")
+
+        if cluster_id == -1:
+            raise ValueError(
+                "Cannot calculate weighted centroid for -1 cluster "
+                "since it is a noise cluster"
+            )
+
+        mask = self.labels_ == cluster_id
+        cluster_data = self._raw_data[mask]
+        cluster_membership_strengths = self.probabilities_[mask]
+
+        if mode == "centroid":
+            return np.average(
+                cluster_data, weights=cluster_membership_strengths, axis=0
+            )
+
+        # mode == "medoid"
+        dist_mat = pairwise_distances(
```

Comment:
So here it might be the only place where we need the distance matrix with no changes. If we have the constructor parameter indicating the need to store the matrix, then we could trigger a copy (I still would need a bit more the code to be certain).

Reply:
But this is certainly here that we want to use the PairwiseDistanceReduction indeed.